keps/sig-node/20201224-pull-image-preference.md

```diff
@@ -0,0 +1,110 @@
+---
+title: image pull priority when serializing image pulls
+authors:
+  - "@saberrey"
+  - "@pacoxu"
+owning-sig: sig-node
+participating-sigs:
+reviewers:
+  - "@odinuge"
+  - "@smarterclayton"
+approvers:
+  - ""
+
+editor: TBD
+creation-date: 2020-12-24
+last-updated: 2020-12-24
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# image pulls priority
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Implementation Details](#implementation-details)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Graduation Criteria](#graduation-criteria)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+When image pull is serialized, container runtime will pull image one by one. If there are some critical pods and others pods on the same server, image pull may start from no critical pods images.
+Hence, we need to make priority for image pull on kubelet.
+
+First of all, --serialize-image-pulls default is true.
+> Pull images one at a time. We recommend *not* changing the default value on nodes that run docker daemon with version < 1.9 or an `aufs` storage backend. Issue #10959 has more details. (DEPRECATED: This parameter should be set via the config file specified by the Kubelet's `--config` flag. See https://kubernetes.io/docs/tasks/administer-cluster/kubelet-config-file/ for more information.)
+
+If --serialize-image-pulls=false is set in your env, that would be great. You can ignore this.
```

Comment:
rather than complicate serialized image pulling behaviors, i would prefer to understand why users are unable to move to parallel-image-pulls since we first had the constraint.  is `aufs` heavily used?  another option is we just always do parallel image pulls as part of cri graduation, and leave it up to the runtime to enforce serial or non-serial semantics.

/cc @mrunalp

Reply:
Most users are using default kubelet configurations. I think that's why we need this.

However, I think the best way is enabling `parallel-image-pulls`.
- Another way to make users use `parallel-image-pulls` by default is to change the default value in kubeadm/kube-spray installing.